test: refactor test that ensures expected versions

coderbyheart opened the pull request.

coderbyheart committed:
test: refactor test that ensures expected versions

MLopezJ commented on adr/008-default-lwm2m-version.md:
I am thinking about it and I think it is an architecture decision, even when there is not much discussion or justification. We are using some versions in a deliberate way and it is a conscious decision. If it will be removed from ADR, I would like to have it very explicit in the readme.  

Here https://github.com/MLopezJ/asset-tracker-lwm2m-js/commit/e72ad32c5d75b02074b304683e8c4f534332d905  was requested to convert the notes in ADR. (https://github.com/MLopezJ/asset-tracker-lwm2m-js/tree/e72ad32c5d75b02074b304683e8c4f534332d905#default-lwm2m-version)

coderbyheart replied:
And now it's part of a test, which makes it even more binding than when it's written down in a README or another Markdown file, because they are not used to drive the tests.

MLopezJ replied:
The test comment was updated with the old ADR 008 information (for a human a table is easier to read than a matrix)

https://github.com/MLopezJ/asset-tracker-lwm2m-js/pull/7/commits/3577796b9793c9cf10600e0bc5a081799daffb6d

MLopezJ commented on src/schemas/lwm2m-types-versions.spec.ts:
```diff
@@ -0,0 +1,36 @@
+	parseURN,
+	Pressure_3323_urn,
+	Temperature_3303_urn,
+} from './index.js'
```
I can do it but just to point out that all the URN  in the project must change to be imported from `/schemas/index.js` and not from `@nordicsemiconductor/lwm2m-types`

coderbyheart replied:
Yes, this is the intention here. To clearly document that we are re-using definitions from another library.

MLopezJ replied:
Imports were updated https://github.com/MLopezJ/asset-tracker-lwm2m-js/pull/7/commits/3bc425bc297bb23d0bafc9f9576d562c9751bc9e

MLopezJ commented on src/schemas/index.ts:
```diff
@@ -0,0 +1,10 @@
+import * as types from '@nordicsemiconductor/lwm2m-types'
+
```
Export `Config_50009` from `index`, this way all URNs can be imported from one single file.

MLopezJ replied:
resolved in https://github.com/MLopezJ/asset-tracker-lwm2m-js/pull/7/commits/3bc425bc297bb23d0bafc9f9576d562c9751bc9e

MLopezJ reviewed (commented):
I can implement the changes I suggest, I just add the comments to open the discussion and hear your point of view there.

coderbyheart commented:
That's not relevant for this specific test, since it does not concern the custom object.

MLopezJ committed:
docs: document expected LwM2M and object versions

MLopezJ committed:
docs: add LwM2M and object version to data transition table

MLopezJ committed:
docs: update test comment

MLopezJ committed:
fix: import LwM2M definitions from local file

URN definitions and types definitions

MLopezJ commented on src/schemas/index.ts:
```diff
@@ -0,0 +1,10 @@
+export const Device_3_urn = types.Device_3_urn
+export const Humidity_3304_urn = types.Humidity_3304_urn
+export const Location_6_urn = types.Location_6_urn
+export const parseURN = types.parseURN
```
`parseURN` function was removed from this file because the intention of this file is to clarify the version of objects used and that function has no relation with that topic. It can be imported from ` @nordicsemiconductor/lwm2m-types` as normal.

MLopezJ replied:
Comments were added to make it even clearer what version is being used. 

Type definitions are now also exported because the URN and the type definition are correlated with the version used.  

https://github.com/MLopezJ/asset-tracker-lwm2m-js/blob/3bc425bc297bb23d0bafc9f9576d562c9751bc9e/src/schemas/index.ts

coderbyheart committed:
test: improve test runtime 10 fold (#10)

Update the test command to improve runtime.

MLopezJ committed:
fix: rename file

remove documents/nRFAssetTracker.md and create documents/data-transition.md instead.

Content of the file is the same.

MLopezJ commented:
WIP: Solving conflicts files

coderbyheart commented on package.json:
```diff
@@ -4,7 +4,7 @@
   "main": "index.js",
   "scripts": {
-    "test": "find ./ -type f -name '*.spec.ts' ! -path '*/node_modules/*' -exec npx tsx {} ';'",
+    "test": "find ./ -type f -name '*.spec.ts' -not -path './node_modules/*' -print | xargs npx tsx --no-warnings --test",
```
Unrelated changed

MLopezJ replied:
`improve-unit-test-time` branch was merged to `it-is-not-a-nonfunction-test` (https://github.com/MLopezJ/asset-tracker-lwm2m-js/pull/10). It means that change hasn't gone into `saga` yet. That's why this change is here and I think it makes sense.

coderbyheart commented on src/schemas/lwm2m-types-versions.spec.ts:
```diff
@@ -0,0 +1,45 @@
+/**
+ * The following versions are expected to be used by the LwM2M objects:
+ *
+ * | Object ID | Object Version | LwM2M Version |
+ * | --------- | -------------- | ------------- |
+ * | 3         | 1.2            | 1.1           |
+ * | 4         | 1.3            | 1.1           |
+ * | 6         | 1.0            | 1.0           |
+ * | 3303      | 1.1            | 1.0           |
+ * | 3304      | 1.1            | 1.0           |
+ * | 3323      | 1.1            | 1.0           |
```
Duplication from https://github.com/MLopezJ/asset-tracker-lwm2m-js/pull/7/files#diff-bb17387fec55dfb8c096118aba859c06009d5e88d0c968615f8a7c3b7be6fcea

MLopezJ replied:
Solved here 49dbbce3b42466ee5e749d7f309cccd16697ec8d

coderbyheart commented on src/schemas/lwm2m-types-versions.spec.ts:
```diff
@@ -0,0 +1,45 @@
+ * | 3304      | 1.1            | 1.0           |
+ * | 3323      | 1.1            | 1.0           |
+ *
+ * Failure to follow these versions could lead to errors.
```
Explain this hypothesis.

MLopezJ replied:
It is difficult to explain every potential issue the failure to follow the recommended versions could create. What I wanted to point out is that I guarantee that with those versions the converter will work but with other versions it could or could not. We already know one case where using different versions of the recommended version creates an unexpected behavior. 

However, by removing the duplication of information, that line was removed as well. 49dbbce3b42466ee5e749d7f309cccd16697ec8d

MLopezJ committed:
fix: refer to documentation to explain expected versions

Avoid duplication of information

MLopezJ committed:
docs: update JSDoc link

refers to local link instead of external
